Phase 1

matajoh opened the pull request:
Adds makeimmutable and isimmutable global functions.

matajoh commented on Include/object.h:
```diff
@@ -125,6 +125,8 @@ check by comparing the reference count field to the immortality reference count.
 #define _Py_IMMORTAL_REFCNT (UINT_MAX >> 2)
 #endif
 
+#define _Py_IMMUTABLE 1
```
Add cast to type

matajoh commented on Include/internal/pycore_object.h:
```diff
@@ -91,6 +93,15 @@ static inline void _Py_ClearImmortal(PyObject *op)
+{
+    if(op) {
+        op->ob_region = _Py_IMMUTABLE;
+        op->ob_refcnt = _Py_IMMORTAL_REFCNT;
```
TODO comment to remove this later

matajoh commented on Python/errors.c:
```diff
@@ -1950,6 +1950,25 @@ PyErr_ProgramTextObject(PyObject *filename, int lineno)
+    if (!_PyErr_Occurred(tstate)) {
+#if __WORDSIZE == 64
+        string = PyUnicode_FromFormat("object is immutable (in region %lu) at %s:%d", obj->ob_region, filename, lineno);
+#else
```
__WORDSIZE == 32

and otherwise error

matajoh commented:
@DavidChisnall

mjp41 commented on Include/internal/pycore_object.h:
```diff
@@ -91,6 +93,17 @@ static inline void _Py_ClearImmortal(PyObject *op)
+    if(op) {
+        op->ob_region = _Py_IMMUTABLE;
+        // TODO once reference counting across regions is fully working
+        // we no longer need to do this
```
```suggestion
        // TODO once reference counting across regions is fully working
        // we no longer need to make all immutable objects immortal
```
I have a habit of coming back to comments like this, and no longer know what I meant by "this".

mjp41 reviewed (approved):
LGTM.  I am not convinced we have all the cases.  I'll start writing some tests on this.

matajoh commented:
Moving this to draft given issues around function objects

davidchisnall commented:
> Moving this to draft given issues around function objects

What is the desired behaviour when you freeze an object that contains a lambda with mutable captures?  Will invoking that lambda then raise exceptions when it's invoked?

matajoh commented:
> > Moving this to draft given issues around function objects
> 
> What is the desired behaviour when you freeze an object that contains a lambda with mutable captures? Will invoking that lambda then raise exceptions when it's invoked?

Freezing the object will make everything it contains immutable. In the case of functions, lambda or otherwise, we introspect into the function and mark the closures, globals, and builtins the function references as immutable. If the lambda is invoked it will only raise exceptions if it modifies the (now immutable) captures. @mjp41  this is what I've done based on my understand of the desired behaviour. Does that track with your understanding?

mjp41 commented on Include/cpython/listobject.h:
```diff
@@ -40,6 +40,12 @@ static inline Py_ssize_t PyList_GET_SIZE(PyObject *op) {
 
 static inline void
 PyList_SET_ITEM(PyObject *op, Py_ssize_t index, PyObject *value) {
+    if(_Py_IsImmutable(op)){
```
You defined a macro `Py_CHECKWRITE`.  Should that be used here?

matajoh replied:
I can't, because it is not currently part of the public API. I may want to move it there, but ultimately we'll replace this code entirely as the Python folks won't want a test in this codepath as it "should only be called by consumers of the C ABI during list init" and it is the only place that needs the check that isn't internal. The real SetItem() has an internal implementation that calls CHECKWRITE appropriately.

mjp41 commented on Include/internal/pycore_veronapy.h:
```diff
@@ -0,0 +1,27 @@
+#include "object.h"
+
+#define Py_CHECKWRITE(op) ((op) && _PyObject_CAST(op)->ob_region != _Py_IMMUTABLE)
+
```
Would a 
```C
#define Py_REQUIREWRITE(op, msg) {if (Py_CHECKWRITE(op)) { _PyObject_ASSERT_FAILED_MSG(op, msg); }}
```
be useful in general?

mjp41 commented on Objects/dictobject.c:
```diff
@@ -1075,7 +1075,7 @@ _Py_dict_lookup(PyDictObject *mp, PyObject *key, Py_hash_t hash, PyObject **valu
             }
             else {
-                *value_addr = DK_UNICODE_ENTRIES(dk)[ix].me_value;
+                *value_addr = _PyDict_EntryValue(DK_UNICODE_ENTRIES(dk) + ix);
```
Not the right spot, but I can't comment on unchanged code.

Should clone preserve immutablity of the keys?

matajoh replied:
Yes. I've fixed this (I believe).

mjp41 commented on Objects/object.c:
```diff
@@ -1172,7 +1172,13 @@ PyObject_SetAttr(PyObject *v, PyObject *name, PyObject *value)
-        err = (*tp->tp_setattro)(v, name, value);
+        if(Py_CHECKWRITE(v)){
+            err = (*tp->tp_setattro)(v, name, value);
+        }else{
```
```suggestion
        }
        else{
```

matajoh replied:
Is this just a style thing? I think the cpython style seems to be open brace on same line. They have some linting they run as a precommit hook and I'm passing that atm.

mjp41 commented on Objects/object.c:
```diff
@@ -1182,7 +1188,14 @@ PyObject_SetAttr(PyObject *v, PyObject *name, PyObject *value)
+
+        if(Py_CHECKWRITE(v)){
+            err = (*tp->tp_setattr)(v, (char *)name_str, value);
+        }else{
```
```suggestion
        }
        else{
```

mjp41 commented on Objects/setobject.c:
```diff
@@ -2347,6 +2390,7 @@ PySet_Pop(PyObject *set)
         PyErr_BadInternalCall();
         return NULL;
     }
+
```
```suggestion
```

mjp41 commented on Python/clinic/bltinmodule.c.h:
```diff
@@ -1409,4 +1409,22 @@ builtin_issubclass(PyObject *module, PyObject *const *args, Py_ssize_t nargs)
+"makeimmutable($module, obj, /)\n"
+"--\n"
+"\n"
+"Make \'obj\' and its entire graph immutable.");
```
```suggestion
"Make \'obj\' and its entire reachable object graph immutable.");
```

mjp41 commented on Python/bltinmodule.c:
```diff
@@ -2737,6 +2739,496 @@ builtin_issubclass_impl(PyObject *module, PyObject *cls,
+    node* head;
+} stack;
+
+stack* stack_new(void){
+    stack* s = (stack*)malloc(sizeof(stack));
+    if(s == NULL){
+        return NULL;
+    }
+
+    s->head = NULL;
+
+    return s;
+}
+
+bool stack_push(stack* s, PyObject* object){
+    node* n = (node*)malloc(sizeof(node));
+    if(n == NULL){
+        return true;
+    }
+
+    n->object = object;
+    Py_INCREF(object);
+    n->next = s->head;
+    s->head = n;
+    return false;
+}
+
+PyObject* stack_pop(stack* s){
+    if(s->head == NULL){
+        return NULL;
+    }
+
+    node* n = s->head;
+    PyObject* object = n->object;
+    s->head = n->next;
+    free(n);
+    Py_DECREF(object);
+
+    return object;
+}
```
I think this should be optimised in a future PR.

matajoh replied:
Agreed.

mjp41 commented on Python/bltinmodule.c:
```diff
@@ -2737,6 +2739,496 @@ builtin_issubclass_impl(PyObject *module, PyObject *cls,
+        _Py_VPYDBGPRINT(element);
+        _Py_VPYDBG("]\n");
+        if(!_Py_IsImmutable(element)){
+            Py_INCREF(element); // element.rc = x + 1
+            if(stack_push(frontier, element)){
+                Py_DECREF(element);
+                return PyErr_NoMemory();
+            }
```
`stack_push` performs an IncRef in the successful case.  Can we write this just as
```suggestion
            if(stack_push(frontier, element)){
                return PyErr_NoMemory();
            }
```

mjp41 commented on Python/bltinmodule.c:
```diff
@@ -2737,6 +2739,496 @@ builtin_issubclass_impl(PyObject *module, PyObject *cls,
+    size = PySequence_Size(seq);
+    _Py_VPYDBG("list/tuple: pushing %ld [i: element] elements\n", size);
+    for(Py_ssize_t i = 0; i < size; i++){
+        PyObject* element = PySequence_GetItem(seq, i); // element.rc = x + 1
+        _Py_VPYDBG("[%ld: ", i);
+        _Py_VPYDBGPRINT(element);
+        _Py_VPYDBG("]\n");
+        if(!_Py_IsImmutable(element)){
+            if(stack_push(frontier, element)){
+                Py_DECREF(element); // element.rc = x
+                return PyErr_NoMemory();
+            }
+        }else{
+            Py_DECREF(element); // element.rc = x
+        }
```
I assume that `PySequence_GetItem` gives an reference count out.  As `stack_push` also does an RC, then this is going to double RC in the successful case, so I think we will have a memory leak.  As we are making everything immutable, that is okay for now.  But when we get to the SCC code, this will be problematic.

mjp41 commented on Python/bltinmodule.c:
```diff
@@ -2737,6 +2739,496 @@ builtin_issubclass_impl(PyObject *module, PyObject *cls,
+            _Py_VPYDBG("<");
+            _Py_VPYDBGPRINT(key);
+            if(!_Py_IsImmutable(key)){
+                Py_INCREF(key); // key.rc = x + 1
+                if(stack_push(frontier, key)){
+                    Py_DECREF(key); // key.rc = x
```
Stack push is already performing the incref that is required here.
```suggestion
                if(stack_push(frontier, key)){
```

mjp41 commented on Python/bltinmodule.c:
```diff
@@ -2737,6 +2739,496 @@ builtin_issubclass_impl(PyObject *module, PyObject *cls,
+            _Py_VPYDBG(": ");
+            _Py_VPYDBGPRINT(value);
+            if(!_Py_IsImmutable(value)){
+                if(stack_push(frontier, value)){
```
I assume `PyObject_GetItem` returns a RC on the object it returns.  So I think we are counting twice in the successful case.

mjp41 commented on Python/bltinmodule.c:
```diff
@@ -2737,6 +2739,496 @@ builtin_issubclass_impl(PyObject *module, PyObject *cls,
+        _Py_VPYDBGPRINT(value);
+        _Py_VPYDBG(")\n");
+
+        if(stack_push(frontier, value)){
```
Same comment on stack_push again

mjp41 commented on Python/bltinmodule.c:
```diff
@@ -2737,6 +2739,496 @@ builtin_issubclass_impl(PyObject *module, PyObject *cls,
+            _Py_VPYDBG(": ");
+
+            if(PyDict_Contains(globals, name)){
+                PyObject* value = PyDict_GetItem(globals, name); // value.rc = x + 1
```
I think the success paths will need to decref value.

mjp41 commented on Python/bltinmodule.c:
```diff
@@ -2737,6 +2739,496 @@ builtin_issubclass_impl(PyObject *module, PyObject *cls,
+                    _Py_VPYDBG("immutable\n");
+                    Py_DECREF(value); // value.rc = x
+                }
+            }else if(PyDict_Contains(builtins, name)){
```
```suggestion
            }
            else if(PyDict_Contains(builtins, name)){
```

mjp41 commented on Python/bltinmodule.c:
```diff
@@ -2737,6 +2739,496 @@ builtin_issubclass_impl(PyObject *module, PyObject *cls,
+                }
+
+                Py_DECREF(value); // value.rc = x
+            }else{
```
```suggestion
            }
            else {
```

mjp41 commented on Python/bltinmodule.c:
```diff
@@ -2737,6 +2739,496 @@ builtin_issubclass_impl(PyObject *module, PyObject *cls,
+            _Py_VPYDBGPRINT(value);
+            _Py_VPYDBG(": ");
+            if(!_Py_IsImmutable(value)){
+                Py_INCREF(value); // value.rc = x + 1
```
As `stack_push` increments in the success cases, I think this logic needs fixing.

mjp41 reviewed (commented):
I think the logic around `stack_push` is assuming it doesn't internally reference count.

matajoh commented on Include/internal/pycore_dict.h:
```diff
@@ -193,6 +203,8 @@ _PyDictValues_AddToInsertionOrder(PyDictValues *values, Py_ssize_t ix)
     *size_ptr = size;
 }
 
+extern int _PyDict_SetItem_Take2(PyDictObject *op, PyObject *key, PyObject *value);
```
Duplicate?

matajoh commented on Include/internal/pycore_runtime_init.h:
```diff
@@ -77,7 +77,7 @@ extern PyTypeObject _PyExc_MemoryError;
                 }, \
                 .context_token_missing = { \
-                    .ob_base = _PyObject_HEAD_INIT(&_PyContextTokenMissing_Type) \
+                    .ob_base = _PyObject_HEAD_INIT(&_PyContextTokenMissing_Type, _Py_IMMUTABLE) \
```
Why are these immutable?

matajoh commented on Objects/regions.c:
```diff
@@ -0,0 +1,537 @@
+    return PyCFunction_Check(obj) || Py_IS_TYPE(obj, &_PyMethodWrapper_Type) || Py_IS_TYPE(obj, &PyWrapperDescr_Type);
+}
+
+// PyObject* walk_sequence(PyObject* seq, stack* frontier) {
```
No need to leave this in

matajoh committed:
Phase 1

This PR makes the changes to add deep immutability as a first-class language construct in Python.
Objects marked as immutable via the new `make_immutable` built-in function are immutable from
that point forward, and this property extends to all objects in their graph as well as their type.
Attempts to edit immutable objects will result in the new `NotWriteableError` being raised.

Signed-off-by: Matthew Johnson <matjoh@microsoft.com>

mjp41 reviewed (approved):
LGTM, let's merge

mjp41 committed:
Bug fix for __dict__access

Accessing a __dict__ causes the representation of an object to change.  This means that it was previously possible to freeze an object, and then update its representation, and the new representation would be mutable.

This fixes that issue and adds testing for it.
